Add comparison macros for floating point numbers (ck_assert_floating_*)

9fcc opened the pull request:
Those macros could help testing geometry functions with specified precision.

9fcc commented:
I'm a newbye to github. Could somebody explain why does AppVeyor fails and how to debug this?

prusnak commented:
```
C:\projects\check\tests\check_check_sub.c(595) : error C2124: divide or mod by zero
```

9fcc commented:
Need some help again. I've figured out how to open logs for ms builds. With nan I'll deal somehow on different compilers. But logs for platform=ms are unclear for me. Can I get tests log somehow with asserts text?

9fcc commented:
I've removed nan, finite and infinite asserts until get managed to check them on any compiler (ignoring nan's sign).

brarcher commented:
Jenkins: ok to test
^ Will start the last bit of the tests

brarcher commented on src/check.h.in:
```diff
@@ -659,6 +660,117 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+ *
+ * @since 0.10.1
+ */
+#define ck_assert_floating_eq(X, Y) _ck_assert_floating(X, ==, Y)
```
You may not want to offer a way to use the == operator on two floating point numbers. Because floating point math is not associative and may follow different rounding models, the outcome of a computation may not be exactly what one expects. The ck_assert_floating_eq_prec() call below is a much better option. I would request that this and the ck_assert_floating_ne() methods be removed.

brarcher commented on tests/check_check_master.c:
```diff
@@ -104,6 +104,22 @@ static master_test_t master_tests[] = {
   { "Simple Tests", CK_FAILURE, "Assertion '1%d >= 3%f' failed: 1%d == 0, 3%f == 1" },
   { "Simple Tests", CK_PASS,    "Passed" },
+  /* Tests on floating macros */
+  { "Simple Tests", CK_FAILURE, "Assertion 'x == y' failed: x == 1.100000, y == 1.200000" },
```
You will need to rebase against master. I added a change today which will add an entry to this table. Now one will need to include the name of the method under test. For example, the first entry will now become:

{ "Simple Tests", "test_ck_assert_floating_eq", CK_FAILURE, "Assertion 'x == y' failed: x == 1.100000, y == 1.200000" }

brarcher commented on src/check.h.in:
```diff
@@ -659,6 +660,117 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+do { \
+  long double _ck_x = (X); \
+  long double _ck_y = (Y); \
+  unsigned char _prec = (PREC); \
```
The PREC parameter currently assumes that one wants to compare the precision of decimal places assuming a power of 1. However, it may be likely that one is interested in comparing using other powers. For example, if I have two floating point numbers: 1e10 and 1.1e10 I may be interested in checking that they are within 1e8 of each other.

Consider changing this to allow one to pass in the maximum value that the two numbers can be apart and still be considered 'equal'.

9fcc replied:
Power of 10 is often used in calculations when assuming fix point calculations which are used in embedded systems. Through fixed point types are unable under C standard using float is much or less a good way.

brarcher replied:
Thanks for the explanation. I now have a better appreciation for where you are trying to go with this macro.

Given your use of the macro, having this to help with math which is expected to be an approximate for fixed point seems reasonable. I still believe that for the floating point macros to be complete a more general macro which allow one to pass the threshold should be provided. Can you provide both in your change?

brarcher commented on tests/check_check_sub.c:
```diff
@@ -443,6 +443,151 @@ START_TEST(test_ck_assert_uint_expr)
+}
+END_TEST
+
+START_TEST(test_ck_assert_floating_eq_prec)
```
When testing this method should data available in [cfloat.h](http://www.cplusplus.com/reference/cfloat/) be used to determine the correctness? This will result in more extreme examples that check boundary conditions.

brarcher commented:
I'm OK considering this change. However, I warn that working with floating point values is hard and determining good and meaningful comparisons is non-trivial. Can you share your expected use case for these macros and the application you are considering using them to unit test?

9fcc commented:
We use those tests in our commercial project so I cannot give full example.  But it looks like this:
ck_assert_double_prec(&speed_integrated, &pos_delta, precision);
We use *_double_* instead of floating to rebase correctly in future.

brarcher commented on src/check.h.in:
```diff
@@ -659,6 +660,117 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+#define _ck_assert_floating(X, OP, Y) do { \
+  long double _ck_x = (X); \
+  long double _ck_y = (Y); \
+  ck_assert_msg(_ck_x OP _ck_y, "Assertion '%s' failed: %s == %.6Lf, %s == %.6Lf", #X" "#OP" "#Y, #X, _ck_x, #Y, _ck_y); \
```
Having no expectation on the values being compared using the format specifier %.6Lf may be a little restrictive. For example, if X were  0.000000012 then the value of 0.000000 would be printed. However, the format %Le would print 1.200000e-08.

Would you be opposed to changing the format specifier to be more general?

9fcc replied:
I thought about it. But tests should give human readable output. So in case of 0.001 and 1.0e-3, 0.001 is clear. But 1.0e-3 spends your time to do some calculations in your mind. In this case %Lg is universal solution.  Just didn't checked what output could be on different compilers and system. In any case I'll make some changes in a week or two to tests making use regular expressions for output analysis (if don't mind) through nan and -nan problem appeared, so no problem, I'll add %Lg everywhere. I can contribute to open source only in free from work time so for now it wouldn't be fast.

brarcher commented:
> Need some help again. I've figured out how to open logs for ms builds. With nan I'll deal somehow on different compilers. But logs for platform=ms are unclear for me. Can I get tests log somehow with asserts text?

Turns out that the Visual Studios build was never printing the output of the unit tests, even on failure. This is now resolved in master ([this pull request](https://github.com/libcheck/check/pull/73)). The Visual Studios build will now always emit the unit test output. Thanks for pointing this out.

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,129 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+  long double _ck_y = (Y); \
+  int _ck_p = (PREC); \
+  int _prec = -_ck_p; \
+  ck_assert_msg(_prec <= 0, \
```
If this assertion is to help protect the usage of this macro consider changing its check to something that relates directly to the user's input. That is,
```
ck_assert_msg((PREC) > 0, ...)
```
This will help a user make sense of the failure.

9fcc replied:
User see custom message after condition and condition (PREC) > 0 won't be seen. How could it help?

brarcher commented:
There are a few failures in the unit tests. One of them I neglected to warn about, sorry about that. In addition to adding the name of the unit test to the master tests table, one will need to add the following call to the unit tests as well:
```
record_test_name(tcase_name());
```

The second failure is here:
/home/travis/build/libcheck/check/tests/check_check_master.c:379:F:Core Tests:test_check_failure_msgs:0: For test 60:Simple Tests:test_ck_assert_floating_eq_prec_min_prec Expected Assertion '-min_prec_minus_one <= 0' failed (prec): min_prec_minus_one == -1, got Assertion '-LDBL_MIN_10_EXP <= 0' failed (prec): LDBL_MIN_10_EXP == -4931

It appears that the test_ck_assert_floating_eq_prec_min_prec unit test failed (as expected) but it did not fail in the way it was expected to fail.

9fcc commented:
My main problem is that
ctest -V -R check_check
goes toooo long because of sleeps. Is there simple any way to run tests by their name or something?

brarcher commented:
There is a way to disable all the timeout tests:
```
./configure --disable-timeout-tests
```
With that you will find that unit tests run very quickly.

9fcc commented:
Well I've done some research. %Lg with default precision is confusing through 0.1 is not 0.1 because of digits ate the end of number. So made macro CK_FLOATING_DIG that can be set by user to determine his own precision (for example DECIMAL_DIG for long double from C99). In documentation I wrote restriction on use of floating point numbers. And I cannot test bounds of precision because of  different representations on x32 and x64 systems (or on different compilers/platforms, didn't dig for the reason).

brarcher commented:
Jenkins: test this please
^ for some reason Jenkins did not re-run the tests on the new change. Forcing it.

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,263 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+do { \
+  int _ck_p = (PREC); \
+  int _prec = -_ck_p; \
+  ck_assert_msg(_prec <= 0, \
```
This is a little confusing because the documentation states P is positive but the checks themselves use -P. I do not see the value in negating PREC in these checks. Could the initial value of PREC be used. This may help the user if an assertion does fail.

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,263 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+ *
+ * @note If the check fails, the remaining of the test is aborted
+ *
+ * @since 0.10.1
```
The next version of Check will be 0.11, so please update the ```@since``` tags

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,263 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+ *
+ * If not P >= 0 or not P <= |LDBL_MIN_10_EXP|, the test fails.
+ *
+ * @param P positive number of decimal places to be compared (precision)
```
This could be simplified to say:
```
Checks that P, used to indicate the precision of a floating point number, is valid.
For P to be valid P>=0 and P<=|LDBL_MIN_10_EXP|
```

As this is not intended to be used directly by a user the ```@note``` and ```@since``` tags are unnecessary. As far as I am concerned this is an 'internal' API.

brarcher replied:
Should this also check
```
P <= LDBL_MAX_10_EXP
```
?

brarcher replied:
The documentation says that ```P<=|LDBL_MIN_10_EXP|```, however the code never performs and abs operation on PREC. What is the purpose of the abs operation in the documentation?

9fcc replied:
ck_assert_floating_prec tests checks for the negative sign so it is equal to abs. But I'll use abs to remove confusing.

9fcc replied:
Documentation says about math used. It could differ from the code if it is easier for understanding and equal to the code result.

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,263 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+ *
+ * @since 0.10.1
+ */
+#define ck_assert_floating_eq_prec(X, Y, PREC) \
```
Could the code in ck_assert_floating_eq_prec and ck_assert_floating_ne_prec be combined and reused? It looks like the only difference is the operation used, either < or >=.

9fcc replied:
Usually I combine code parts if they could be reused 3 or more times. For 2 times I analize amount of resulting code. In this case there are no coins exept further refactor which probably wouldn't be needed. But ok, I'll combine this.

brarcher commented on tests/check_check_sub.c:
```diff
@@ -497,6 +497,284 @@ START_TEST(test_ck_assert_uint_expr)
+}
+END_TEST
+
+START_TEST(test_ck_assert_floating_lt_prec)
```
The gt and lt tests show that the macro correctly fails when x<y or x>y, respectively, however it does not check for failure on the == case. Maybe also show this.

brarcher commented on tests/check_check_sub.c:
```diff
@@ -497,6 +497,284 @@ START_TEST(test_ck_assert_uint_expr)
+
+  int max_prec = -LDBL_MIN_10_EXP;
+  ck_assert_floating_eq_prec(0, 0, max_prec);
+  // TODO: test with -(LDBL_MIN_10_EXP - 1) and platform specific output.
```
Why not use LDBL_MAX_10_EXP to test for the maximum allowed precision?

9fcc replied:
Well, meaning of precision is number of digits after decimal point. So We pass this number to all functions. But it is negative power of ten. So we need to test exactly against LDBL_MIN_10_EXP. Testing against LDBL_MAX_10_EXP would mean that I could use ten in positive power. But it has no sense in context of digits after decimal point.

9fcc commented on src/check.h.in:
```diff
@@ -669,6 +671,263 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+ *
+ * @since 0.10.1
+ */
+#define ck_assert_floating_prec(PREC) \
```
Well, I just didn't decide would user need this. But for now it seems not. So ok.

9fcc commented:
Now I doubt about `prec` suffix. Need some advice about naming. It must indicate what argument used last. As it is error equal to power of 10 I could use `e10` suffix (like in functions log10, exp10). Or `ee10` meaning `error exponent 10`. Thus later I could make `e` suffix that would indicate that user can pass his own error value. An so on.

Next advice I need is may I move all the floating macros in a separate header file? Since their amount would be significant in a time it could reduce amount of conflicts in merges in future. But newbies won't find locations of macros so fast as they could. So I don't know what is better.

brarcher commented:
> Need some advice about naming.

Consider if there were a generic method which compared two floating point numbers within a given error margin. The error could also be called a tolerance. A more specific method is being proposed where the error margin/tolerance is instead 10^{-n}. May I propose something like: _tol10. That is, compare within a tolerance, where the tolerance is a power of 10. That is similar to your idea, but leaves open the possibility of adding _tol macros later where the error is passed in directly. What do you think?

> Next advice I need is may I move all the floating macros in a separate header file?

I do not understand your concern about "in a time it could reduce amount of conflicts in merges in future.". Can you explain further?

I think keeping the macros in the same file is preferable. This way a user needs only include one file to get all the macros. If the floating point macros were in another header file also pulled in by check.h it would only add a level of indirection.

9fcc commented:
> I do not understand your concern about "in a time it could reduce amount of conflicts in merges in future.". Can you explain further?
Well when project actively developing there is a big chance that two people may edit the same file. If branches are separate from each other with long time pulling one branch into another becomes a headache. Git remembers changes you made by last merging. But if one line changed by two users it can show conflict in whole file (against parent commit of both branches). The easiest way we resolved the issue is making as much file as possible. The problem disappeared.

> tolerance
Well it is the same as error. Error is more suitable in context of floating point numbers. It's probably standard de facto in pointing to difference between real value and calculated value. I was confused when I didn't found another used word but it is so.

9fcc commented:
> tolerance

Sorry you were right. Error means difference between real and calculated values but acceptable difference is tolerance. I'll rename those macros.

9fcc commented:
Done. Since now name points to power of ten I removed some mess in the code. So now I check full bounds with LDBL_MAX_10_EXP and precision would be negative in meaning of decimal places.

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,246 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+ * Check two floating point numbers to determine if not X≈Y with specified
+ * tolerance
+ *
+ * If X ≈ Y with error < powl(10, -P), the test fails.
```
Oh, never mind, you are correct.

brarcher commented on tests/check_check_sub.c:
```diff
@@ -497,6 +497,337 @@ START_TEST(test_ck_assert_uint_expr)
+
+  int min_tol10 = LDBL_MIN_10_EXP;
+  ck_assert_floating_eq_tol10(0, 0, min_tol10);
+  // TODO: test with LDBL_MAX_10_EXP+1 and platform specific output.
```
Is this TODO necessary?

brarcher replied:
Or, rather, should it be testing that if LDBL_MIN_10_EXP-1 is passed that it fails?

9fcc replied:
I've fixed TODO. As I cannot test platform dependent error messages until I add regular expressions to check_check_master.c I think TODO is necessary.

brarcher replied:
Oh, I see. As you do not know what the maximum or minimum number is offline, and certainly the value may be different between different platforms, you cannot know what the failure message will be.

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,51 @@ respectively.
+@itemx ck_assert_floating_ge
+
+Compares two floating point numbers (@code{float}, @code{double},
+@code{long double}) and displays predefined message with condition and
```
...and displays a predefined message with both the condition and input parameters on failure.

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,51 @@ respectively.
+to @code{==}, @code{!=}, @code{<}, @code{<=}, @code{>}, and @code{>=}
+respectively with acceptable error (tolerance) specified by the last parameter.
+Beware of using those operators for floating operators because of unclear bounds
+of exponent and because floating point errors are platform dependant. Floating
```
dependant => dependent

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,51 @@ respectively.
+point numbers often internally represented as binary so they can't be exactly
+power of 10. All these operators have significant error in comparison so use
+them only if you know what you're doing. As axample
+@code{ck_assert_floating_le_tol10(0.002*10, 0.001*10, 2)} would fail
```
Is the assumption about platform really an assumption that some platforms may not conform to IEEE 754, the IEEE Standard for Floating-Point Arithmetic standard? If two platforms conform to this, would it not be expected that the the result is the same?

9fcc replied:
Here is the answer:
[http://stackoverflow.com/questions/18494237/floating-point-mismatch-between-compilers-visual-studio-2010-and-gcc](http://stackoverflow.com/questions/18494237/floating-point-mismatch-between-compilers-visual-studio-2010-and-gcc)
I've figured it out on your ms platform tests while testing bounds. GCC and clang were passed. But one of compilers didn't passed test. So I had to remove testing bounds.

Moreover I''ll remove ge_tol10 and le_tol10. They could not give the same result on some values on different platforms and in fact they break logic of comparing with specified precision because values 0.01 and 0.02 could become equal with tolerance pow(10, -2) in case of <=. With gt operator they wouldn't.

9fcc replied:
Must correct myself. I'll remove gt_tol10 and lt_tol10. Tolerance includes equal value.

brarcher replied:
I did not realize that there could be so much variation in the implementation of floating point math between different compilers. That is good to know.

9fcc commented on doc/check.texi:
```diff
@@ -1032,6 +1032,51 @@ respectively.
+@code{<}, @code{<=}, @code{>}, and @code{>=} respectively. Beware
+using those operators for floating point numbers because of precision
+possible loss on every operation on floating point numbers. For example
+(1/3)*3==1 would return false, because 1/3==1.(3) and cannot be represented
```
https://en.wikipedia.org/wiki/Repeating_decimal
It turns out that different countries have different notations of repeating decimal. I had never knew about ellipsis notation. In my country braces are used. Thanks for pointing this.

brarcher replied:
Oh, I did not realize that either. Your change where you write both notations is good. Thanks for clearing this up.

9fcc commented:
I've made about 1000+ lines of tests for floating point numbers in the project I work for and can confirm that specially prepared algorithms can be tested with predefined precision level.

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,73 @@ respectively.
+
+Compares two floating point numbers (@code{float}, @code{double},
+@code{long double}) with specified user tolerance set by the third
+parameter (@code{unsigned char}) and displays predefined message with condition
```
The third parameter is a long double, right? The tol10 macros use a unsigned char third parameter .

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,73 @@ respectively.
+error (tolerance) specified by the last parameter.
+Beware of using those operators for floating operators because of unclear bounds
+of floating point numbers and because floating point errors are platform
+dependent. Floating point numbers often internally represented as binary
```
... Floating point numbers are often internally represented in binary so they cannot be exact powers of 10.

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,73 @@ respectively.
+@code{eq}, @code{ne}, @code{le}, and @code{ge} correspond to @code{==},
+@code{!=}, @code{<=}, and @code{>=} respectively with acceptable
+error (tolerance) specified by the last parameter.
+Beware of using those operators for floating operators because of unclear bounds
```
"Beware using these functions for floating point comparisons because of (1) unclear bounds of floating point numbers, and (2) floating point numerical errors are platform dependent."

Can you explain "unclear bounds of floating point numbers"? I do not follow.

9fcc replied:
I mean that 0.1 is not exactly 0.1. There would be digits after some zeroes in decimal places. So in fact number would be a little greater than 10.

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,73 @@ respectively.
+Some assertions could fail on one platform and would be passed on another.
+For example expression @code{0.02<=0.01+10^-2} is true by meaning,
+but some platforms may calculate it as false. IEEE 754 standard specifies
+representation of floating point numbers format but it does not promise that
```
The IEEE 754 standard specifies the floating point number format representation but it does not promise that the same computation carried out on all hardware will produce the same result.

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,73 @@ respectively.
+@code{eq}, @code{ne}, @code{le}, and @code{ge} correspond to @code{==},
+@code{!=}, @code{<=}, and @code{>=} respectively with acceptable
+error (tolerance) specified by the last parameter.
+Beware of using those operators for floating operators because of reasons same
```
Beware of using these functions for floating point comparisons due to the same issues mentioned for the floating point tolerance comparisons mentioned earlier.

brarcher commented:
My only comments are documentation now. I think that these macros have come along really well. I appreciate your effort!

You may also want to update the AUTHORS file to mention these macros.

9fcc commented:
I've updated AUTHORS file and added some more asserts (checking for NaN and for infinity). I've started using INFINITY constant for a certain situations so soon I may need those macros too.

I had to add TODOs for nonnan tests because NaN values could be unsigned and signed. Some compilers prints "nan" and some other "-nan" (but the sign is irrelevant).

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,396 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+#define ck_assert_floating_le_tol10(X, Y, P) _ck_assert_floating_op_tol10(X, <, Y, P, 1)
+
+/**
+ * Check floating point number to be finite (none of those: +infinity,
```
Check that a floating point number is finite; i.e. is not +infinity, -infinity, or NaN.

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,396 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+} while (0)
+
+/**
+ * Check floating point number to be infinite (+infinity or -infinity)
```
Check that a floating point number is infinite, either +infinity or -infinity.

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,396 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+} while (0)
+
+/**
+ * Check floating point number to be NaN
```
Check that a floating point number is not a number (NaN).

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,396 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+ *
+ * @since 0.11.0
+ */
+#define ck_assert_floating_nonnan(X) \
```
Is there another way to say ~isnan()? Does "real" make sense?

9fcc replied:
I've just made as others do writing nonnull. It a special value such as the null constant. Real is just another name of float type in another languages (like Pascal). So such naming could be confusing. May be something like _correct/_incorrect? Since NaN indicates error in calculations.

brarcher replied:
I could not find an alternative that mades sense. nonnan is fine.

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,99 @@ respectively.
+
+@item ck_assert_floating_finite
+
+Checks floating point number (@code{float}, @code{double}, @code{long double})
```
Checks that a floating point number (...) is finite and ...

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,99 @@ respectively.
+
+@item ck_assert_floating_infinite
+
+Checks floating point number (@code{float}, @code{double}, @code{long double})
```
Checks that a floating point number (...) is infinite and ..

brarcher commented on doc/check.texi:
```diff
@@ -1032,6 +1032,99 @@ respectively.
+
+@item ck_assert_floating_nan
+
+Checks floating point number (@code{float}, @code{double}, @code{long double})
```
Checks that a floating point number (...) is NaN (not a number) and

9fcc commented on AUTHORS:
```diff
@@ -8,60 +8,61 @@ Ex-Maintainers:
     Sven Neumann      <sven at convergence dot de>                   |  neo23
 
 Committers:
-    Robert Collins        (subunit support)                          |  rbcollins
```
I'll split it into two commits. But can you explain why is that necessary?

brarcher replied:
> But can you explain why is that necessary?

It help with code archeology if commits that change only space are separate from commits that change code. In this case it is not that important. However, in general I try to have space-only changes in their own commits.

> Wouldn't it be better to split authors and their role into different lines? Role could be indented relative to an author.

If you would like to make this change, that would be fine. It is your call.

9fcc replied:
Done.

9fcc commented:
I've changed `not a number` to "Not a Number". This way manual page for isnan describes it.

brarcher commented on tests/check_check_sub.c:
```diff
@@ -497,6 +497,511 @@ START_TEST(test_ck_assert_uint_expr)
+  double t = 1;
+  double x = 0;
+  ck_assert_floating_nonnan(x);
+  // TODO: test with 0.0/0.0 and platform specific output.
```
What could be done so that these could be tested? You mentioned that NaN could be negative or positive. Is it possible to ensure that if the value is NaN that it is positive before printing it out? Or, is it simply that some platforms will always print NaN is negative?

9fcc commented:
I found situation when I need to double tolerance of tol10 precision. Making point fixed goes doubling error for some variables. So in some situations I have `tol < 10^-p` and in others  `tol < 2 * 10^-p` (round error in line equation). It is correct but tol10 macros are not suitable for this (I used tol macros for now). So I confused about them. I want to add something like coefficient to the power of 10. So it would be kind a `k * pow(10, -p)`.
Changing macros such a way would add new situations for use but macros would loss their meaning and would become complicated.
Adding new macros would increase overhead. And I can't think of a suffix for those new macros.

What do you think about it?

brarcher commented:
Perhaps only keep the _tol macros and remove the _tol10 macros. That way one may compute whatever tolerance one wants and use the _tol macros, and Check can remain more simple. Do you believe that this would suit your use case?

9fcc commented:
Well, It would be better if Check would show original variables without computations. At the moment I test only one precision level but soon I'll add testing intervals of precision. So it would be one of the input parameters. Precision printed as giant floating point number would produce much more extra output. I want to make Check some kind a good floating point testing framework. tol macros only show the result but not the initial parameters of tolerance.

Removing tol10 would transfer pow(10, -p) and 2 * pow(10, -p) into the tests making code more complex. So I'll have to make wraps for those checks and I'll come to tol10 macros in any case.

tol10 could be marked as experimental. It would mean that if it would have much critics it would be removed.

9fcc commented:
You've already said your opinion about splitting header file into parts. But what about some kind of plug-in system? We could put into check.h only basic macros. And all experimental or case specific macros and functions we can put into different header files. If users needs them he includes appropriate file. In future we can analyse usage of experimental macros and move some of them into main file.

Why so? Because distributors could compile Check without unwanted macros. And they could be easily thrown away in next versions just removing files.
What do you think about it?

9fcc commented:
I removed tol10 from this merge request. But probably will add them in some other merge request. So changes are almost ready for merge. What about "unclear bound"? Any suggestions? Or keep as is?

9fcc commented:
Realized one more problem that could exists. Converting from float to long double (at most computers is double) produces number that is not equal to the original. So macros need some rewriting (I'll remove _ck_* variables). As I used double in tests they passed. But they may not on float.

9fcc commented:
Didn't found any good solution without using C11 so I've splitted macros into three groups. Need some final review.

brarcher commented on src/check.h.in:
```diff
@@ -669,6 +671,696 @@ CK_DLL_EXP void CK_EXPORT _ck_assert_failed(const char *file, int line,
+ *
+ * @since 0.11.0
+ */
+#define ck_assert_float_nonnan(X) _ck_assert_floating_nonnan(X, double, "l")
```
should this be ck_assert_double_nonnan?

brarcher commented:
> Realized one more problem that could exists. Converting from float to long double (at most computers is double) produces number that is not equal to the original. So macros need some rewriting (I'll remove ck* variables). As I used double in tests they passed. But they may not on float.

You have observed that changing two floats to long doubles, then comparing them, may end up as a different result?

9fcc commented:
It's more complicated. For example, numbers 0.1f, 0.1 and 0.1l are not equal. So if someone would test float variable but second argument would be of double type (just number) the result would be erroneous.
That is because converting from one floating point type to another has loss of precision. For example if we convert float to double we would have the same 7 digits of precision. But double have 15 bits of precision. So they would show garbage.
[http://stackoverflow.com/questions/3498192/c-convert-double-to-float-preserving-decimal-point-precision](http://stackoverflow.com/questions/3498192/c-convert-double-to-float-preserving-decimal-point-precision)

So to prevent situations when user doesn't understand what is happening I added macros for every type.

brarcher commented:
> For example, numbers 0.1f, 0.1 and 0.1l are not equal. So if someone would test float variable but second argument would be of double type (just number) the result would be erroneous.
That is because converting from one floating point type to another has loss of precision. For example if we convert float to double we would have the same 7 digits of precision. But double have 15 bits of precision. So they would show garbage.

This would apply for converting from a type with more precision (long double) to one with less precision (float). However, if one always converts to long double how would this result in an issue? Is there evidence that converting from float to long double results in 'garbage' (i.e. undefined behavior)?

9fcc commented:
Here is the example:
	printf("%g; %lg; %d\n", 0.1f, 0.1, 0.1f == 0.1);
// 0.1; 0.1; 0
// As shows printf they are equal, but...
	printf("%.15g; %.15lg; %d\n", 0.1f, 0.1, 0.1f == 0.1);
// 0.100000001490116; 0.1; 0
    volatile float x = 0.1f;
    volatile double y = 0.1;
    printf("%.15lg; %.15lg; %d\n", (double) x, y, ((double) x) == y);
// 0.100000001490116; 0.1; 0
// That was happening when converting float to double.

In fact float has much more than 7 decimal points. But 7 decimal can be set to desired values. All other are garbage. And converting to longer type you're just moving that garbage into precision decimal points. That's all the magic.

Although converting from one type to another will give you closest possible value. I didn't checked but it could bring much more undefined situations. For example when you expect something like 0.2 (0.20000000125...) you will get 0.199999999999999985.... And it could be closer to 0.2 than previous value. But that is only my assumption.

brarcher commented:
Having so many macros for different floating point comparisons looks rather cumbersome. Let me ask a few more questions to see if there is no other way.

The macros first store the two values being compared, then perform the computation with the local variables. Because of this, a type must be used. I do not see how this is a hard requirement. What would happen if, in the example of _ck_assert_floating_op_tol, the following was tried:
```
int _result = (X - Y) OP T * (D);
ck_assert_msg(_result, ...
```
Could this avoid knowing the exact type, leaving it up to whatever the caller specified them as?

> And converting to longer type you're just moving that garbage into precision decimal points. That's all the magic.
> Although converting from one type to another will give you closest possible value. I didn't checked but it could bring much more undefined situations.

Consider if one cast a float to a long double. If the original value only had a precision of 7.25 decimal places, one would expect that the long double value to have no more precision than that. The remaining digits would be expected to be noise. A user should not expect to get any additional precision by casting up. The case where it would be troublesome is if casting up resulted in a loss of precision.

9fcc commented:
> What would happen if, in the example of _ck_assert_floating_op_tol, the following was tried:

It was the first suggestion I tried to implement. But I had to reject this idea because I cannot get type of variables to apply appropriate modifiers to `fprintf` (i.e. `%f`, `%lf`, `%Lf`). The only way I can get type of macro argument is the sizeof. But sizeof(int) == sizeof(float). So it could be segmentation fault in some cases. In other it would be type mismatch and wrong results.

9fcc commented:
> Consider if one cast a float to a long double. If the original value only had a precision of 7.25 decimal places, one would expect that the long double value to have no more precision than that. The remaining digits would be expected to be noise. A user should not expect to get any additional precision by casting up. The case where it would be troublesome is if casting up resulted in a loss of precision.

I think is  a very bad idea. User could pass first argument as float and second argument as double. But converting one to another will get new numbers. The are not the same. If a macro if only for float numbers user would not be misunderstood. he already knows that float numbers must be passed.

9fcc commented:
One more example:
```
    double d = 0.1;
    float f = 0.1;
    float fl = 0.1l; // in fact it would be 0.1f
    printf("%.15lf; %.15f; %.15f\n", d, f, fl);
// 0.100000000000000; 0.100000001490116; 0.100000001490116
```

So in case of typed macros constants passed to a macro would be converted to variable type. Thus it resolves the problem when user passes 0.1 comparing to float variable. It would be just converted to 0.1f. But not promoted. So they would be equal.
Here is the explanation:
[http://stackoverflow.com/questions/36840173/precision-loss-from-float-to-double-and-from-double-to-float](http://stackoverflow.com/questions/36840173/precision-loss-from-float-to-double-and-from-double-to-float)

brarcher commented:
(Reasoning about floating point math is not easy...)

The link you mentioned, namely,

http://stackoverflow.com/questions/36840173/precision-loss-from-float-to-double-and-from-double-to-float

mentions that promoting a float->{double, long double} or double->{long double} does not change the value. If this is the case, does that not imply that the following:

> In fact float has much more than 7 decimal points. But 7 decimal can be set to desired values. All other are garbage. And converting to longer type you're just moving that garbage into precision decimal points. That's all the magic.

is not entirely accurate, as the lower order bits of the [significand field](https://en.wikipedia.org/wiki/Floating_point#Internal_representation) would be '0'. The following [example](http://ideone.com/Lt10lV) may show this:
```
    volatile float x = 0.1f;
    volatile double y = 0.1;
    printf("%.15lg; %.15lg; %.15g;\n", (double) x, y, x);
    // Prints: 0.100000001490116; 0.1; 0.100000001490116;
```
That is, the float variable 'x' when cast to a double and printed, and printed directly, is the same value.

> > Consider if one cast a float to a long double.

> I think is a very bad idea. User could pass first argument as float and second argument as double. But promoting one to another will produce new digits in precision zone. The are not the same.

True, but my understanding is that the new digits would be '0' and would not change the value of the data. Thus, if a user did decide to compare a float and a double, but the double happen to exactly represent the same data at the float, then the comparison would still be valid. If, however, they were not expected to be exact, then the tolerance-based macros would be appropriate, no?

I'm still leaning on the side of having one macro that casts to long double then compares. Using separate macros may be pedantic, but it may not be strictly necessary. Do you disagree?

> UPDATE: I thought about code size. If we generate three files from one .h.in file, it would decrease code amount.

I'm not too concerned about the size of check.h. If a user does not use a macro in a given file then it should not impact their code size.

9fcc commented:
> That is, the float variable 'x' when cast to a double and printed, and printed directly, is the same value.

Well, yes, they are the same for computer but it still may confuse a programmer who is trying to solve simple task and doesn't know what floating point is, because in decimal representation they are not the same. And it's true only for promotions. Example:

```
	volatile float x = 0.1;
	volatile double y = 0.1;
	fprintf(stderr, "%.15f %.15lf %d\n", x, y, (x == y));
//0.100000001490116 0.100000000000000 0
```

> I'm not too concerned about the size of check.h. If a user does not use a macro in a given file then it should not impact their code size.

I mean the code size to maintain.

> I'm still leaning on the side of having one macro that casts to long double then compares. Using separate macros may be pedantic, but it may not be strictly necessary. Do you disagree?

If I thought if isn't necessary I wouldn't rewrite all the code once again. The example is in the tests. I added one more test to show what I mean (test_ck_assert_ldouble_eq_with_conv).

9fcc commented:
And in any case float, double and long double separate macros are just give users a chance to choose. One could use ldouble macros for all types. Another does not need double and long double type. So he can use float macros. And float macros should be much faster than double (or ldouble) macros in i586 architecture. For example my tests for only one library already run for about a second. And it's only beginning.

9fcc commented:
I've added tests for expressions.

brarcher commented:
> The example is in the tests. I added one more test to show what I mean (test_ck_assert_ldouble_eq_with_conv).

```
float x = 1.1;
long double y = x;
ck_assert_ldouble_eq(x, y);
record_failure_line_num(__LINE__);
ck_assert_ldouble_eq(x, 1.1);
```

I think that I understand now. At first glance one would have expected that to pass. But, the 1.1 being passed to the macro was originally a long double, were as the float was promoted, means that unless 1.1 can be exactly represented as a float then the comparison will fail. Tricky.

Reviewing the changes one more time, I am comfortable with these additions. Thanks for the effort in adding them, and thanks for the discussion. I'll update the branch, and after the checks complete I'll merge the changes in.

The last piece necessary for this work is adding platform dependent checks for NaN, right? Any thoughts on how best to achieve that in a maintainable way?

9fcc commented:
> The last piece necessary for this work is adding platform dependent checks for NaN, right? Any thoughts on how best to achieve that in a maintainable way?

I'll add regular expressions to the master file. I think it can be useful when adding new API. So NaN would be just "-?nan" in the text. Another way, I could add a check to the macro and print always nan as text. But I am not sure about second solution because I do not know what user expects to see. If he always see -nan while debugging and tests would show +nan, it could confuse a little. But it's a much easier way.
